### PR TITLE
Uncertainties and correlations

### PR DESCRIPTION
**Pull request by whaeck**
This adds the capability to calculate the uncertainties and correlations for covariance blocks.

Uncertainties can only be calculated for covariance blocks that are on the diagonal in the covariance matrix. The function that calculates the uncertainties therefore has no effect when called on a covariance block that is off diagonal in the covariance matrix.

Uncertainties are required to convert the covariances into correlations. For a covariance block on the diagonal of the covariance matrix, this can be done without external input (the uncertainties can be calculated from the covariances). For an off diagonal covariance block, they cannot be calculated without input of the uncertainties. As a result, there are two functions to calculate the correlations: one for diagonal and one for off diagonal blocks.

The matrix() method to retrieve the covariance matrix of a block was renamed to covariances().

### CODE REVIEW

File: src/codex/CovarianceBlock.hpp [hunk 1]
```diff
@@ -13,17 +13,24 @@ namespace codex {
...
+    /* fields - meta data */
     MetaData row_;
     std::optional< MetaData > column_;
-    Matrix< double > matrix_;
+
+    /* fields - covariance matrix */
+    std::optional< Matrix< double > > covariances_;
+
+    /* fields - uncertainties and correlations */
+    std::optional< std::vector< double > > uncertainties_;
```
Comment (ptalou) [inline src/codex/CovarianceBlock.hpp:32]: Maybe replace 'uncertainties' by 'variances' vs 'standard_deviations'

File: src/codex/CovarianceBlock/src/ctor.hpp [hunk 2]
```diff
@@ -6,9 +6,10 @@
  */
 CovarianceBlock( MetaData metadata, Matrix< double > matrix ) :
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/ctor.hpp:7]: rename matrix as covariances

File: src/codex/CovarianceBlock/src/calculateCorrelations.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ *  The correlations can only be calculated with input of the uncertainties
+ *  for covariance blocks that are off diagonal in the covariance matrix.
+ *  Uncertainties will not be stored.
+ *
+ *  @param[in] row      the uncertainties to be applied to each row
+ *  @param[in] column   the uncertainties to be applied to each column
+ */
+void calculateCorrelations( const std::vector< double >& row,
+                            const std::vector< double >& column ) {
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/calculateCorrelations.hpp:37]: Please rename row and column to be more explicit, i.e., uncertainty-related.

Comment (ptalou) [reply src/codex/CovarianceBlock/src/calculateCorrelations.hpp:37]: Should you include a test of the possible size mismatch between row/column and numberGroups?

**Review by ptalou — CHANGES_REQUESTED** [review]

File: src/codex/CovarianceBlock.hpp (hunk 1 above)
Comment (whaeck) [reply src/codex/CovarianceBlock.hpp:32]: I picked standard deviations as the term to go with. If we want the variances specifically (i.e. the square of the standard deviations), we can just add a function to do so.

**Comment by whaeck** [comment]
Changes were added to update/reviews

**Review by ptalou — APPROVED** [review]

**Pull request merged**